File: App-13/app/src/main/java/com/android_lab/app_13/MainActivity.java
package com.android_lab.app_13;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.Manifest;
import android.content.ContentResolver;
import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Telephony;
import android.widget.ArrayAdapter;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.Toast;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity implements MyAdapter.ClickListener {

    private static final int REQUEST_READ_SMS = 01;
    RecyclerView recyclerView;
    private MyAdapter adapter;
    private LinearLayoutManager layoutManager;
    private ArrayList<Message> sms  = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        recyclerView = findViewById(R.id.recyclerView);
        layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);

        adapter = new MyAdapter(sms,this);
        recyclerView.setAdapter(adapter);

        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(this, layoutManager.getOrientation());
        recyclerView.addItemDecoration(dividerItemDecoration);

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_SMS) == PackageManager.PERMISSION_GRANTED) {
            readMessages();
        } else {
            requestPermission();
        }
    }

    private void requestPermission() {
        if(ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.READ_SMS)) {

        } else {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.READ_SMS}, REQUEST_READ_SMS);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode) {
            case REQUEST_READ_SMS:
                if(grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED)
                    readMessages();
                else
                    Toast.makeText(this, "Permission Denied", Toast.LENGTH_LONG).show();
                return;
                
            default:
                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }

    private void readMessages() {
        ContentResolver contentResolver = getContentResolver();
        Cursor c = contentResolver.query(Uri.parse("content://sms/inbox"), null, null, null, null);

        if (((c != null) ? c.getCount() : 0) > 0) {
            while (c != null && c.moveToNext()) {
                String address = c.getString(c.getColumnIndex(Telephony.Sms.ADDRESS));
                String body =  c.getString(c.getColumnIndex(Telephony.Sms.BODY));
                sms.add(new Message(address, body));
            }
        } else {
            sms.add(new Message("No Message on Your Device", ""));
        }

        if (c != null) {
            c.close();
        }

        adapter.dataset = sms;
        adapter.notifyDataSetChanged();
    }

    @Override
    public void onClick(int position) {
        Message m = sms.get(position);

        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage(m.body)
                .setTitle(m.address+"\n")
                .setNeutralButton("Done", new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int id) {
                            dialog.dismiss();
                        }
                    });
        builder.create().show();
    }
}

File: App-11_kotlin/app/src/main/java/com/android_lab/app_11_kotlin/MainActivity.kt
package com.android_lab.app_11_kotlin

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Toast
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.json.JSONArray
import java.io.File
import java.io.InputStreamReader

class MainActivity : AppCompatActivity(), MyAdapter.ClickListener {

    private lateinit var recyclerView: RecyclerView
    private lateinit var viewAdapter: RecyclerView.Adapter<*>
    private lateinit var viewManager: RecyclerView.LayoutManager

    private var cars = arrayOf<Car>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        cars = readJSON()

        viewManager = LinearLayoutManager(this)
        viewAdapter = MyAdapter(cars, this)
        val itemDecoration = DividerItemDecoration(this, (viewManager as LinearLayoutManager).getOrientation())

        recyclerView = findViewById<RecyclerView>(R.id.recyclerView).apply {
            setHasFixedSize(true)
            layoutManager = viewManager
            adapter = viewAdapter
            addItemDecoration(itemDecoration)
        }
    }

    override fun onClick(position: Int) {
        Toast.makeText(this, "Selected: " + cars.get(position), Toast.LENGTH_LONG).show()
    }

    fun readJSON(): Array<Car> {
        val json = assets.open("cars.json").bufferedReader().use { it.readText() }
        return Gson().fromJson(json, Array<Car>::class.java)
    }

}
